src/ThematicBreak.elm

```diff
@@ -15,108 +15,71 @@ type ThematicBreak
 
-                    ( Finished, Hyphen occurrences ) ->
-                        succeedIfEnough occurrences
 
-                    ( Finished, Underscore occurrences ) ->
-                        succeedIfEnough occurrences
-
-                    ( Star, Asterisk occurrences ) ->
-                        Advanced.Loop (Asterisk (occurrences + 1))
-                            |> succeed
-
-                    ( Dash, Hyphen occurrences ) ->
-                        Advanced.Loop (Hyphen (occurrences + 1))
-                            |> succeed
-
-                    ( TokenUnderscore, Underscore occurrences ) ->
-                        Advanced.Loop (Underscore (occurrences + 1))
-                            |> succeed
-
-                    ( TokenUnderscore, NoMatchYet ) ->
-                        Advanced.Loop (Underscore 1)
-                            |> succeed
+parseThematicBreak : Parser ThematicBreak
+parseThematicBreak =
+    oneOf
+        [ withChar '-'
+        , withChar '*'
+        , withChar '_'
+        ]
 
-                    ( Star, NoMatchYet ) ->
-                        Advanced.Loop (Asterisk 1)
-                            |> succeed
 
-                    ( Dash, NoMatchYet ) ->
-                        Advanced.Loop (Hyphen 1)
-                            |> succeed
+{-| Per the commonmark spec:
 
-                    ( Whitespace, _ ) ->
-                        Advanced.Loop state
-                            |> succeed
+> a sequence of three or more matching -, \_, or \* characters,
+> each followed optionally by any number of spaces, forms a horizontal rule.
 
-                    _ ->
-                        problem (Parser.Expecting (stateToString state))
-            )
+-}
+withChar : Char -> Parser ThematicBreak
+withChar tchar =
+    let
+        token =
+            tokenHelp (String.fromChar tchar)
+    in
+    succeed ThematicBreak
+        |. token
+        |. whitespace
+        |. token
+        |. whitespace
+        |. token
+        |. chompWhile (\c -> c == tchar || isSpace c)
+        |. oneOf
+            [ tokenHelp "\n"
+            , end (Parser.Expecting "end")
+            ]
 
 
-succeedIfEnough occurences =
-    if occurences > 2 then
-        succeed (Done ThematicBreak)
+isSpace : Char -> Bool
+isSpace c =
+    case c of
+        ' ' ->
+            True
 
-    else
-        problem (Parser.Expecting "...?")
+        '\t' ->
+            True
 
+        _ ->
+            False
 
-stateToString : State -> String
-stateToString state =
-    case state of
-        Asterisk v ->
-            "Asterisk " ++ String.fromInt v
 
-        Hyphen v ->
-            "Hyphen " ++ String.fromInt v
+whitespace : Parser ()
+whitespace =
+    chompWhile isSpace
 
-        Underscore v ->
-            "Underscore " ++ String.fromInt v
 
-        NoMatchYet ->
-            "NoMatchYet"
+singleSpace : Parser ()
+singleSpace =
+    chompIf isSpace (Parser.Expecting "Space")
```

Comment:
these parsers are extracted because now they are shared. If defined inline, a program could re-define this value many times, using space each time.